pkg/zsio/zeek/flatten.go

```diff
@@ -0,0 +1,75 @@
+package zeek
+
+import (
+	"fmt"
+
+	"github.com/mccanne/zq/pkg/zeek"
+	"github.com/mccanne/zq/pkg/zson"
+	"github.com/mccanne/zq/pkg/zson/resolver"
+	"github.com/mccanne/zq/pkg/zval"
+)
+
+type Flattener struct {
+	mapper *resolver.Mapper
+}
+
+func NewFlatener() *Flattener {
+	return &Flattener{
+		mapper: resolver.NewMapper(resolver.NewTable()),
+	}
+}
+
+func (f *Flattener) Flatten(r *zson.Record) (*zson.Record, error) {
+	id := r.Descriptor.ID
+	d := f.mapper.Map(id)
+	if d == nil {
+		cols := flattenColumns(r.Type.Columns)
+		outRecord := zeek.LookupTypeRecord(cols)
+		d = f.mapper.Enter(id, outRecord)
+	}
+	if d.Type == r.Descriptor.Type {
+		r.Descriptor = d
+		return r, nil
+	}
+	// XXX this loop should build a native zval
+	var ss []string
+	it := r.ZvalIter()
+	for _, col := range r.Descriptor.Type.Columns {
+		val, isContainer, err := it.Next()
+		if err != nil {
+			return nil, err
+		}
+		recType, isRecord := col.Type.(*zeek.TypeRecord)
+		if isRecord {
+			it2 := zval.Iter(val)
```

Comment:
will this do the right thing if the container is unset?

Reply:
You tell me.  It's your code.  No it doesn't work.

Reply:
If we're talking about the search api code in "something else", then as discussed on the PR, that was a stopgap since there's no current way to ingest an unset record into "something else" and the impending switch to raw payloads in that api means that doing work on the search api code would be pointless.